message/validation/common_checks.go

```diff
@@ -140,3 +140,16 @@ func (mv *messageValidator) validateBeaconDuty(
 
 	return nil
 }
+
+func (mv *messageValidator) signerIndexInCommittee(signer spectypes.OperatorID, committee []spectypes.OperatorID) int {
+	// Probably converting committee to a map would be faster, but since committee size is <13, this should be okay.
+	// Although, we need to change it if it appears in pprof output.
+	for i, oid := range committee {
+		if oid == signer {
+			return i
+		}
+	}
+
+	mv.logger.Panic(fmt.Sprintf("signer %v must be in committee %v", signer, committee))
+	panic("unreachable") // fix compilation issue
```

Comment:
@y0sher 

This must be checked in advance by message validation, so triggering `mv.logger.Panic` would mean a bug in message validation, so I think this must be left as a panic. 

Regarding the second panic, we usually prefer `mv.logger.Panic` over `panic` but the go compiler doesn't know that `mv.logger.Panic` always panics and doesn't allow it to compile. However, now I think that just `panic` should be enough because it's unreachable

Reply:
I'd also probably go with `error` (not `panic`) in this case, 

I don't think it's OK to `panic` here because here we have many "actors" and it takes only 1 of them to deny service to all the rest (even though panic here would mean we have a bug, we probably don't want to let "bad" actors punish "good" actors if we do - for example, after bad code-merge we accidentally remove 1 validation code line).

In other words, it's generally bad to violate the following principle: "independent components/services/handlers/actors should function independently (meaning failure in 1 of them shouldn't affect others)".

Reply:
@iurii-ssv 

> I don't think it's OK to panic here because here we have many "actors" and it takes only 1 of them to deny service to all the rest (even though panic here would mean we have a bug, we probably don't want to let "bad" actors punish "good" actors if we do - for example, after bad code-merge we accidentally remove 1 validation code line).

IMO accidental removal of 1 line is a critical bug, which may lead to much worse consequences than the ability to DDoS. I think we need to be able to spot this bug ASAP. If we change it to error, no one would notice it

Reply:
> IMO accidental removal of 1 line is a critical bug, which may lead to much worse consequences than the ability to DDoS. I think we need to be able to spot this bug ASAP. If we change it to error, no one would notice it

Hm, yes that's a trade-off, 

but it seems to me that 
- any "legitimate" message that triggers this error/panic (which can happen if we have that missing-line bug) should/will be spotted by unit/spec tests we have - I guess we can run them and check
- and any "illegitimate" message ... we don't really care about immediately spotting these

Reply:
> Also, there are a lot of other assumptions based on previous checks in message validation, and if we add all of them, it will inflate the code a lot.

That's a good point, indeed we can't cover them all - so the question is whether we want to try and cover those that are "easily visible" (like the one we are discussing here).

> But it's impossible for any message to trigger this, it's against message validation rules. 

I understand that,

> Moreover, now all errors are covered with tests, and I want to keep this enforced, however, it's impossible to write a test for this case if it's an error because it's unreachable

Right, what I meant is to 1) comment out the validation we have (the one that prevents us from reaching this `panic`) and 2) run tests to see if they discover this panic (if they do, I think it's "safe/fine" to replace it with error).

That's a lot of effort though, I'm not sure myself it's worth it. Perhaps @y0sher has anything to add on this.

Reply:
Now this part looks like this:
```go
func (ci *CommitteeInfo) signerIndex(signer spectypes.OperatorID) int {
	idx, ok := ci.signerIndices[signer]
	if !ok {
		panic(fmt.Sprintf("BUG: message validation must have checked that signer %v is in committee %v", signer, ci.committee))
	}

	return idx
}
```
And I tend to change it to the below (or even remove this method) to keep it consistent with other message validation code (we don't usually check invariants) because I don't like adding an error

```go
func (ci *CommitteeInfo) signerIndex(signer spectypes.OperatorID) int {
	return ci.signerIndices[signer]
}
```